### PR TITLE
 added code for reading network interface

### PR DESCRIPTION
**Pull request by NeethuVisw**

### CODE REVIEW

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 1]
```diff
@@ -93,6 +93,42 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
     return err;
 }
 
+CHIP_ERROR DiagnosticDataProviderImpl::GetNetworkInterfaces(NetworkInterface ** netifpp)
+{
+    struct netif * net_interface;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    NetworkInterface * ifp = new NetworkInterface();
+
+    net_interface = netif_find("en0");
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: we should remember to add a task to the wifi story to extend this with both/all.

Comment (MauritsFassaert) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Done: https://sensorfy.atlassian.net/browse/TSBM-236

Comment (n-nievergeld) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Didn't you find the classes to iterate over all lwip interfaces?

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 2]
```diff
@@ -93,6 +93,42 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
     return err;
 }
 
+CHIP_ERROR DiagnosticDataProviderImpl::GetNetworkInterfaces(NetworkInterface ** netifpp)
+{
+    struct netif * net_interface;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    NetworkInterface * ifp = new NetworkInterface();
```
Comment (n-nievergeld) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Are you sure that this gets only called once?

How about a static global somewhere (or an array?)

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 3]
```diff
@@ -93,6 +93,42 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+
+    net_interface = netif_find("en0");
+    if (net_interface)
+    {
+        ifp->name = CharSpan::fromCharString(net_interface->name);
+        ifp->isOperational = net_interface->flags & NETIF_FLAG_LINK_UP;
+        ifp->type = EMBER_ZCL_INTERFACE_TYPE_ENUM_ETHERNET;
+        ifp->offPremiseServicesReachableIPv4.SetNull();
+        ifp->offPremiseServicesReachableIPv6.SetNull();
+        ifp->hardwareAddress = ByteSpan(net_interface->hwaddr, net_interface->hwaddr_len);
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: I'm still missing the ip addresses?

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 2 above)
Comment (MauritsFassaert) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: I think they are deleted in ReleaseNetworkInterfaces. Making them static would reserve ram permanently for a structure that may be read once or never during the lifetime of the bridge. 

The dynamic memory allocation of course also has a cost (slightly more cpu cost, small chance of fragemntation) but I think it's negligible.

**Review by MauritsFassaert — CHANGES_REQUESTED** [review]
Add IP addresses?

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 3 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: It is updated for Ipv6address, ipv4 is disabled in the config.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 1 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Yes, I've updated code with Inet::InterfaceIterator

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 4]
```diff
@@ -23,6 +23,8 @@
 #include <platform/internal/CHIPDeviceLayerInternal.h>
 
 #include <platform/DiagnosticDataProvider.h>
+#include <app/data-model/List.h>
+#include <lib/support/logging/CHIPLogging.h>
```
Comment (n-nievergeld) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Is this header needed? I did not see that you did add logs.

**Review by n-nievergeld — APPROVED** [review]

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 5]
```diff
@@ -93,6 +95,85 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        }
+
+        NetworkInterface *ifp = new NetworkInterface();
+        if (ifp == nullptr)
+        {
+            err = CHIP_ERROR_NO_MEMORY;
+            break;
+        }
+
+        ifp->name = CharSpan::fromCharString(net_interface->name);
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: isn't this the same as interfaceName ? 
It may not be a big thing but I'm trying to get rid of the net_interface all together. It bypasses a layer of abstraction, which is not the best software design

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 6]
```diff
@@ -93,6 +95,85 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        ifp->isOperational = (net_interface->flags & NETIF_FLAG_LINK_UP) != 0;
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: This could also be `interfaceIterator.IsUp()` which uses the abstraction, both avoiding bypassing the layer and the code duplication/implementation detail of the lower level interface.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 7]
```diff
@@ -93,6 +95,85 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        ifp->type = EMBER_ZCL_INTERFACE_TYPE_ENUM_ETHERNET;
+        ifp->offPremiseServicesReachableIPv4.SetNull();
+        ifp->offPremiseServicesReachableIPv6.SetNull();
+        ifp->hardwareAddress = ByteSpan(net_interface->hwaddr, net_interface->hwaddr_len);
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Also here, I would prefer to use interfaceIterator.GetHardwareAddress(...). With that the net_interface is no longer needed.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 8]
```diff
@@ -93,6 +95,85 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        NetworkInterface *ifp = new NetworkInterface();
+        if (ifp == nullptr)
+        {
+            err = CHIP_ERROR_NO_MEMORY;
+            break;
+        }
+
+        ifp->name = CharSpan::fromCharString(net_interface->name);
+        ifp->isOperational = (net_interface->flags & NETIF_FLAG_LINK_UP) != 0;
+        ifp->type = EMBER_ZCL_INTERFACE_TYPE_ENUM_ETHERNET;
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: How do you know this? It could also be wifi now that we're iterating. 
Luckily there's InterfaceIterator::GetInterfaceType(...) the diagnostics data provider of zephyr at src/platform/Zephyr/DiagnosticDataProviderImpl.cpp#256 shows how. In fact that whole function is basically what we need, including error handling.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 6 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: I've updated

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 5 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Yes, I've updated.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 4 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: I've updated.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 8 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: I've updated InterfaceIterator::GetInterfaceType for renesas platform

File: src/inet/InetInterface.cpp [hunk 9]
```diff
@@ -255,7 +255,21 @@ bool InterfaceIterator::HasBroadcastAddress()
 
 CHIP_ERROR InterfaceIterator::GetInterfaceType(InterfaceType & type)
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    VerifyOrReturnError(HasCurrent(), CHIP_ERROR_INCORRECT_STATE);
+
+    if(!mCurNetif)
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    if (mCurNetif->flags & NETIF_FLAG_ETHARP)
```
Comment (MauritsFassaert) [inline src/inet/InetInterface.cpp]: You can use `netif_is_flag_set` and it should be `NETIF_FLAG_ETHERNET`

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 10]
```diff
@@ -93,6 +94,118 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    NetworkInterface *head = nullptr;
+
+    for (Inet::InterfaceIterator interfaceIterator; interfaceIterator.HasCurrent(); interfaceIterator.Next())
+    {
+        NetworkInterface *ifp = new NetworkInterface();
+        interfaceIterator.GetInterfaceName(ifp->Name, Inet::InterfaceId::kMaxIfNameLength);
+        ifp->name          = CharSpan::fromCharString(ifp->Name);
+        if (ifp == nullptr)
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: This should be moved up to before the name now or it will crash before checking for memory errors

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 11]
```diff
@@ -93,6 +94,118 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        {
+            err = CHIP_ERROR_NO_MEMORY;
+            break;
+        }
+
+        ifp->name = CharSpan::fromCharString(ifp->Name);
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: this is done twice

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 12]
```diff
@@ -93,6 +94,118 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+                ifp->type = EMBER_ZCL_INTERFACE_TYPE_ENUM_THREAD;
+                break;
+            case Inet::InterfaceType::Cellular:
+                ifp->type = EMBER_ZCL_INTERFACE_TYPE_ENUM_CELLULAR;
+                break;
+            }
+        }
+        else
+        {
+            ChipLogError(DeviceLayer, "Failed to get interface type");
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: This leaves the ifp->type return value undefined, which is a bug. I see other errors are also handled that way. 
I think not failing entirely (ie returning the error) is perhaps an ok policy, but leaving random enums that could be meaningless would be a bug, since no error is given on the matter interface.

TLDR: `ifp->type = EMBER_ZCL_INTERFACE_TYPE_ENUM_UNSPECIFIED`

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 13]
```diff
@@ -93,6 +94,118 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        }
+
+        ifp->offPremiseServicesReachableIPv4.SetNull();
+        ifp->offPremiseServicesReachableIPv6.SetNull();
+
+        uint8_t addressSize;
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: it's safer to initialize this to 0

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 14]
```diff
@@ -93,6 +94,118 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+
+        err = interfaceIterator.GetHardwareAddress(ifp->MacAddress, addressSize, sizeof(ifp->MacAddress));
+
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(DeviceLayer, "Failed to get network hardware address");
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: I assume the span is default initialized to an empty span? that seems ok then, otherwise make sure the result is defined also in case of warnings. 

Also (not blocking) but having the error case handled first, especially since above it's the other way around, is a bit confusing. I prefer happy flow first unless using an early out.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp [hunk 15]
```diff
@@ -93,6 +94,118 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetBootReason(BootReasonType & bootReason
...
+        }
+        else
+        {
+            ifp->hardwareAddress = ByteSpan(ifp->MacAddress, addressSize);
+        }
+
+        size_t ipv6AddressCount = 0;
+
+        Inet::InterfaceAddressIterator interfaceAddressIterator;
+        while (interfaceAddressIterator.HasCurrent())
```
Comment (MauritsFassaert) [inline src/platform/renesas/DiagnosticDataProviderImpl.cpp]: can be easier to read as a for-loop instead of a while loop. It keeps it all on one line. To me if the decision is made to iterate in a while loop then I assume there's something peculiar going on with the incrementing (eg the list is modified while looping). That doesn't seem to be the case here.

**Review by MauritsFassaert — CHANGES_REQUESTED** [review]
Required:
1) nullptr dereference if out of memory 
2) undefined return value when failing to get the type
the rest of the comments are readability improvement suggestions.

File: src/inet/InetInterface.cpp (hunk 9 above)
Comment (NeethuVisw) [reply src/inet/InetInterface.cpp]: I've updated this, but it is also not working with NETIF_FLAG_ETHERNET

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 15 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: updated

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 14 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: updated

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 12 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: Updated , if it is unknown ifp->type set to EMBER_ZCL_INTERFACE_TYPE_ENUM_UNSPECIFIED in else

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 11 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: removed.

File: src/platform/renesas/DiagnosticDataProviderImpl.cpp (hunk 10 above)
Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: updated

Comment (MauritsFassaert) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: not far enough. in line 105 `ifp` is dereferenced :boom:

Comment (NeethuVisw) [reply src/platform/renesas/DiagnosticDataProviderImpl.cpp]: updated

**Review by MauritsFassaert — APPROVED** [review]
